src/main/kotlin/cacophony/codegen/instructions/CacophonyInstructionCovering.kt

```diff
@@ -22,18 +22,29 @@ class CacophonyInstructionCovering(private val instructionMatcher: InstructionMa
 
     private fun coverWithInstructionsForValue(node: CFGNode, register: Register): List<Instruction> {
         val matches = instructionMatcher.findMatchesForValue(node, register)
-        val bestMatch = matches.maxByOrNull { match -> match.size } ?: error("No match found for $node, ${node.javaClass}")
+        val bestMatch =
+            matches.maxByOrNull { match -> match.size } ?: error("No match with instructions for value found for $node, ${node.javaClass}")
         return coverGivenMatch(bestMatch)
     }
 
     override fun coverWithInstructions(node: CFGNode): List<Instruction> {
         val matches =
-            instructionMatcher.findMatchesForSideEffects(node)
+            instructionMatcher
+                .findMatchesForSideEffects(node)
                 .union(instructionMatcher.findMatchesForValue(node, Register.VirtualRegister()))
         val bestMatch = matches.maxByOrNull { match -> match.size } ?: error("No match found for $node, ${node.javaClass}")
         return coverGivenMatch(bestMatch)
     }
 
+    override fun coverWithInstructionsWithoutVirtualRegisters(node: CFGNode): List<Instruction> {
+        val matches = instructionMatcher.findMatchesForSideEffects(node)
+        val bestMatch =
+            matches
+                .filter { it.toFill.isEmpty() }
+                .maxByOrNull { match -> match.size } ?: error("No match without virtual registers found for $node, ${node.javaClass}")
+        return coverGivenMatch(bestMatch)
+    }
+
     override fun coverWithInstructionsAndJump(node: CFGNode, label: BlockLabel, jumpIf: Boolean): List<Instruction> {
         val matches = instructionMatcher.findMatchesForCondition(node, label, jumpIf)
         val bestMatch = matches.maxByOrNull { match -> match.size } ?: error("No match found for $node, ${node.javaClass}")
```

Comment:
We can also change this error's message as in methods above.

Reply:
done